### ISSUE TITLE
Segmentation fault on -rmTrans 1 command

### ISSUE THREAD

**Issue by nfsaremi** [ANGSD/angsd#238]
I have attempted to use both angsd version 0.928 and 0.929 to create a fasta file from a bam. I would like to remove transitions from the output since my sample is ancient. I am able to run the -doFasta 1 command with '-rmTrans 0', but the program always seg faults when I run '-rmTrans 1'. I have tried running each scaffold separately as well, but no luck. 

angsd -i realigned.bam -doFasta 1 -doCounts 1 -minMapQ 30 -minQ 25 -C 50 -rmTrans 1 -ref ref.fasta

**Comment by ANGSD** [comment]
Dear @nfsaremi , this sounds like a bug. Would it be possible for me to obtain a copy of your bamfile. Or could you try to run the command through a debugger after you have compiled with debug parameters.

make clean;make FLAGS=-ggdb

gdb angsd
run angsd -i realigned.bam -doFasta 1 -doCounts 1 -minMapQ 30 -minQ 25 -C 50 -rmTrans 1 -ref ref.fasta

when it crashes press bt followed by enter.

Best

**Comment by nfsaremi** [comment]
Thanks for getting back to me. I compiled ANGSD again and ran the command through the debugger. Here is the error, and the backtrace output:


Error:
Charnge chr:181
	-> [writeChr] writing chr:scaffold_8258
	-> Printing at chr: scaffold_5561 pos:406163 chunknumber 1000 contains 863 sites
Program received signal SIGSEGV, Segmentation fault.
0x0000000000462793 in abcWriteFasta::run (this=0xda0ce0, pars=0x1024ce0) at abcWriteFasta.cpp:242
242	      int rb = refToInt[pars->ref[pars->posi[s]]];

backtrace:
#0  0x0000000000462793 in abcWriteFasta::run (this=0xda0ce0, pars=0x1024ce0) at abcWriteFasta.cpp:242
#1  0x00000000004c14de in main_analysis (p=0x1024ce0) at shared.cpp:195
#2  0x00000000004c150e in interface (p=0x1024ce0) at shared.cpp:205
#3  0x00000000004c1a40 in selector (p=0x1024ce0) at shared.cpp:338
#4  0x000000000048a749 in callBack_bambi (fff=0x10a4ea0) at angsd.cpp:52
#5  0x00000000004a993f in uppile (show=0, nThreads=1, rd=0xda2b20, nLines=50, nFiles=1, regions=std::vector of length 0, capacity 0, gf=
    0xd9e350) at mUpPile.cpp:1575
#6  0x000000000048d81e in bammer_main (args=0xbf4d20) at bammer_main.cpp:238
#7  0x00000000004b0694 in multiReader::fetch (this=0xbf4c80) at multiReader.cpp:567
#8  0x000000000048b255 in main (argc=19, argv=0x7fffffffe128) at angsd.cpp:230


Let me know if there is anything else I can do to help. Thanks!

**Comment by ANGSD** [comment]
Ok, that helps, could it be that you have have reads that are positioned
after the end of the reference?


Den man. 19. aug. 2019 kl. 19.29 skrev Nedda F Saremi <
notifications@github.com>:

**Comment by nfsaremi** [comment]
That's possible considering my reference is a few million years diverged from the ancient species. Is there any fix you could recommend?

**Comment by avershinina** [comment]
Hello,
I have exactly the same problem with all my BAM files. 
Debug report:
```
	-> angsd version: 0.931 (htslib: 1.9) build(Aug 29 2019 11:21:34)
	-> No '-out' argument given, output files will be called 'angsdput'
	-> Reading fasta: ref.fasta
[bammer_main] 1 samples in 1 input files
	-> Parsing 1 number of samples 
[shared.cpp.changeChr():281] refid:0
Charnge chr:0
	-> Printing at chr: chr10 pos:201592 chunknumber 400 contains 272 sites
Program received signal SIGSEGV, Segmentation fault.
0x00005555555b7294 in abcWriteFasta::run (this=0x555555d6da40, pars=0x555555de7a00) at abcWriteFasta.cpp:242
242	      int rb = refToInt[pars->ref[pars->posi[s]]];
```

Backtrace:
```
#0  0x00005555555b7294 in abcWriteFasta::run (this=0x555555d6da40, pars=0x555555de7a00) at abcWriteFasta.cpp:242
#1  0x000055555560bc1f in main_analysis (p=0x555555de7a00) at shared.cpp:195
#2  0x000055555560bc4f in interface (p=0x555555de7a00) at shared.cpp:205
#3  0x000055555560c1c5 in selector (p=0x555555de7a00) at shared.cpp:338
#4  0x00005555555d3aa4 in callBack_bambi (fff=0x555555ee65b0) at angsd.cpp:52
#5  0x00005555555f3d67 in uppile (show=0, nThreads=1, rd=0x555555d5b500, nLines=50, nFiles=1, 
    regions=std::vector of length 0, capacity 0, gf=0x555555d585e0) at mUpPile.cpp:1575
#6  0x00005555555d6cdd in bammer_main (args=0x555555d57010) at bammer_main.cpp:238
#7  0x00005555555fad31 in multiReader::fetch (this=0x555555d56f70) at multiReader.cpp:567
#8  0x00005555555d462f in main (argc=17, argv=0x7fffffffdc98) at angsd.cpp:230
```

I looked up the region in tview and it does not look problematic.
I also have noticed that various filter settings produce segfailt in different places of the genome. 
For example, if I change -C 50 to -C40, "-> Printing at chr:" shows another position.

**Comment by ANGSD** [comment]
Dear @avershinina and @nfsaremi , what I believe the issue is, is that your reference lengths is shorter than your sequencing data.

Say your fastafile is only 100bp long, but you have  read that starts at position 90. Then the read will extend. Could you verify that this is not the problem?

**Comment by ANGSD** [comment]
I am closing this issue. But feel free to reopen if my answer didnt help.

Best

**Comment by avershinina** [comment]
Dear @ANGSD ,

I think I solved the problem. It looks like not the reads extending to out-of-reference bounds were the problem. I tried Picard SAM cleaner and various tools filtering for soft- and hard-clipped reads, and that did not help. I also tried running angsd on a set of regions within the contigs, with coordinates of these regions at a significant distance away from the ends and starts of the contigs. That did not help either. 

**What worked:**

What helped is filtering BAM for unmapped reads with `samtools view -F4`. I was not aware there were unplaced reads in my alignment, and this filtering did the trick: rmTrans worked without hiccups. 

I can't re-open the issue because @nfsaremi has opened it, but I hope this piece of info will help someone in the future.  

best
Alisa

**Comment by mcassatt** [comment]
@ANGSD I'm also having this same issue - I tried the solution that @avershinina posted and unfortunately wasn't able to get it to run. If I run `-doFasta -rmTrans 0` it can run no problem, but once I include `-rmTrans 1` the program seg faults and dumps. I've tried running it on different chromosomes, as well as on a region within a chromosome well away from either of the ends, so it shouldn't be any out of bounds reads that are causing it to seg fault. I tried running it on the whole assembly and chromosome by chromosome and nothing works.
I used the same input files to run `-doAbbababa 1 -rmTrans 1 ` and it works fine. Any help appreciated!
